ui/migration.md

```diff
@@ -0,0 +1,50 @@
+# Proposal to move the existing UI code to a separate repo
+
+## Current situation
+
+The existing UI in the `visualization` directory is published to `npm` as
+`fairlearn-dashboard` and consumed in the `fairlearn.widget` module.
+Users can view the Fairlearn dashboard through Python code in a Jupyter
+notebook.
+
+## Problems
+
+1. The current setup with both Python and Typescript code in the same
+   repository has lead to confusion for developers who assumed that changes
+   to the visualization code would automatically be reflected in the widget.
+2. Several members of the community have voiced concerns with the dashboard
+   living alongside the "core" Python code. These included:
+   - dashboard is written in Typescript, not Python
+   - dashboard has lots of dependencies that are unrelated to Python
+   - dashboard shouldn't be part of Fairlearn at all
+   - even if it's part of Fairlearn, it should be a dependency and not live in
+     the same repo
+3. The dashboard was built with much more complex requirements in mind than
+   perhaps required for just the open source package. This is due to its use
+   within Azure Machine Learning Studio.
+4. The Python code has been much more open to contributions than the
+   Typescript code. The Typescript code lacks documentation and contributor
+   guide.
+5. Several people mentioned the need to have context-specific visualizations
+   as opposed to a fixed dashboard. From a project/website perspective this
+   would allow us to think of appropriate visualizations per example notebook
+   as opposed to providing a general purpose dashboard with the package.
+
+## Proposal
+
+To address the concerns outlined above we propose to
+
+1. Remove the `visualization` code from Fairlearn.
+2. Remove the `fairlearn.widget` module from the `fairlearn` package.
+3. Move these pieces into a public repository under the `microsoft`
+   organization on GitHub (with MIT license as before).
+   This is to better signal that the package needs to abide by somewhat
+   stricter Microsoft cloud integration requirements.
+4. Publish a package (Name TBD) from the repo in the `microsoft` organization
+   that can be installed separately if desired.
+
```

Comment:
If MS folks on the Azure side aren't able to collaborate or contribute in the open, and decide to create a separate **microsoft/azure-fairlearn** repo that does not have an open development or governance process, that seems like a missed opportunity to collaborate, but it's already the status quo so it seems fine to me to formalize that :)

One specific suggestion is to update this proposal so that it specifies how new development starts within the Fairlearn open source project.  This could build on the values expressed in [Contributing example notebooks](https://fairlearn.github.io/contributor_guide/contributing_example_notebooks.html), the draft [Mission & Roadmap](https://github.com/fairlearn/fairlearn/pull/500), and ideas expressed in [Brainstorming: Website for "fairness as a sociotechnical challenge" community](https://github.com/fairlearn/fairlearn/issues/498) and the [Fairlearn: Theory of change](https://hackmd.io/EBh01XPtRZGHEg76oD1rGw) hackpad.  I'd be happy to try contributing to this work within **fairlearn/fairlearn**!

If folks are interested in collaborating on that, some tangible first steps would be:
1. Update the [quick start example](https://fairlearn.github.io/quickstart.html) so that it demonstrates how to meet the same user needs without using any Azure bits (eg, with calls to matplotlib as @adrinjalali suggested).  Add a link to **microsoft/azure-fairlearn**.
2. Ask Azure staff to move any issues in **fairlearn/fairlearn** that are specifically related to **microsoft/azure-fairlearn** into that repo.
3. Revise the **fairlearn/fairlearn** roadmap to be [centered on user needs and project values](https://github.com/fairlearn/fairlearn/pull/500#discussion_r443040820), and update that prioritization given the level of MS staffing and commitment from open source volunteers.  In that conversation, I'd vote to prioritize fixing immediate issues for new users (eg, examples don't work because published packages lag the website; core assessment features work the first time across all notebook platforms; usage guides aimed at first-time users), new work that starts by expressing how it's aligned with practitioner needs (eg, [Holstein et al. 2019](http://www.jennwv.com/papers/fairness-in-practice.pdf); [Madaio et al. 2020](http://www.jennwv.com/papers/checklists.pdf)), and new work on tools that support engaging with fairness as a sociotechnical challenge (eg, [Selbst et al. 2019](https://andrewselbst.files.wordpress.com/2019/10/selbst-et-al-fairness-and-abstraction-in-sociotechnical-systems.pdf); [candidate screening example notebook](https://hackmd.io/GMli82s7SxORABkabCgw8Q), updating  the website to include [sociotechnical talking points](https://hackmd.io/nDiDafJ6TMKi2cYDHnujtA)).

I'd love to see work that meets practitioner needs, and work that approaches fairness as a sociotechnical challenge as a core part of **fairlearn/fairlearn**.  Thanks for listening! 👍

Reply:
As you're writing this was the status quo as it is, so this isn't a step backwards but rather a formal manifestation of what was happening already. Contributions have never really been possible, but the steps outlined in the proposal will change that by
- specifying how/where to contribute
- specifying the development process
- explaining exactly what restrictions there are (related to localization, accessibility, styling)

That should in fact make it much easier to contribute. As mentioned earlier this is still open on GitHub with MIT license. From what I understand about typescript so far it's possible to create different kinds of wizards from there, e.g. one that AzureML uses and another one that we build based on a candidate screening scenario, and another one that we build based on a lending scenario, etc. Depending on their purpose they can use different views that can live alongside each other in the repo and perhaps also some shared ones.

If people feel like it's easier to do this outside of the Microsoft repo that's perfectly fine, too! As you can read in the updated proposal there are pretty much three ways of setting this up, and whatever works for people can be done. This isn't mutually exclusive with the original proposal, it's rather beyond the original proposal's scope (and that's why that scope is now extended).

The point you made (and @koaning in a different comment) about quickstart, notebooks, and user guides makes sense. I've explicitly added that to the proposed steps.

Regarding your proposed steps: these are all existing items that everyone is more than welcome to join in, and some of them are actively in progress (e.g. @riedgar-ms is versioning the documentation to avoid the difference in the webpage to the last published package version).

The entire point of this proposal is to allow everyone to move beyond restricting development so that UI development can actually happen. If that sounds like an interesting part of Fairlearn to contribute to (in the context of the examples we discuss on Thursdays or otherwise) we'd love to work with you, whether it's in the Microsoft repo or a repo under the Fairlearn org.